feat(treesitter): add `@injection.filename`

clason opened the pull request:
Problem: Injecting languages for file redirects (e.g., in bash) is not possible.

Solution: Add `@injection.filename` capture that is piped through `vim.filetype.match({ filename = node_text })`; the resulting filetype (if not `nil`) is then resolved as a language (either directly or through the list maintained via `vim.treesitter.language.register()`). Also add a sample injection for bash.

Note: `@injection.filename` is a non-standard capture introduced by Helix; having two editors implement it makes it likely to be upstreamed.

Supersedes #27832

@lewis6991

lucario387 commented on runtime/queries/bash/injections.scm:
```diff
@@ -1,3 +1,11 @@
+  body: (command
+    argument: (string
+      (string_content) @injection.content))
+  redirect: (file_redirect
+    destination: (word) @injection.filename)
```
There should be a limit to whether it's `>` or `<`, or something else.

lucario387 replied:
Just to be sure

```bash
echo "foo" > bar < bar.lua
```

```
program [0, 0] - [1, 0]
  redirected_statement [0, 0] - [0, 26]
    body: command [0, 0] - [0, 10]
      name: command_name [0, 0] - [0, 4]
        word [0, 0] - [0, 4]
      argument: string [0, 5] - [0, 10]
        string_content [0, 6] - [0, 9]
    redirect: file_redirect [0, 11] - [0, 16]
      destination: word [0, 13] - [0, 16]
    redirect: file_redirect [0, 17] - [0, 26]
      destination: word [0, 19] - [0, 26]
```

clason replied:
As I don't bash, please make a concrete suggestion.

lucario387 replied:
I don't bash as well but I believe this can lead to duplicated injections (my example above) so it's better to limit it to one stream only (probably stdout?)

lucario387 replied:
Probably something like the below would be better. @lewis6991 
```suggestion
  redirect: (file_redirect
    ">"
    destination: (word) @injection.filename)
```

lewis6991 replied:
I'm not really sure that this is a big problem. Maybe just add a todo?

ObserverOfTime replied:
>     echo "foo" > bar < bar.lua

Given that doesn't work I don't think we need to worry about it.

lewis6991 commented:
These are the injections I am actually using:

```scheme
; echo "
; local a = 1
; " > file.lua
(redirected_statement
  body: (command
    name: (command_name (word) @_cmd)
    argument: [(string) (raw_string)] @injection.content
  )
  redirect: (file_redirect
    destination: (_) @injection.filename
  )
  (#eq? @_cmd echo)
  (#offset! @injection.content 0 1 0 -1)
  (#set! injection.include-children)
)

; cat <<'EOF' > file.lua
; local a = 1
; EOF
(redirected_statement
  body: (_) @_cmd
  redirect: (heredoc_redirect
    redirect: (file_redirect
      destination: (_) @injection.filename
    )
    (heredoc_body) @injection.content
  )
  (#eq? @_cmd cat)
  (#set! injection.include-children)
)
```

lewis6991 commented:
Semi-related but I also have this:

```scheme
(
 (command
   name: (command_name) @command
   redirect: (herestring_redirect
               (string) @injection.content
               )
   )
 (#eq? @command "python3")
 (#offset! @injection.content 0 1 0 -1)
 (#set! injection.language "python")
 (#set! injection.include-children)
 )
```

Which makes me think something like `injection.command` would be better than `injection.shebang`.

clason commented:
> These are the injections I am actually using:
> 
> ```scheme
> ; echo "
> ; local a = 1
> ; " > file.lua
> (redirected_statement
>   body: (command
>     name: (command_name (word) @_cmd)
>     argument: [(string) (raw_string)] @injection.content
>   )
>   redirect: (file_redirect
>     destination: (_) @injection.filename
>   )
>   (#eq? @_cmd echo)
>   (#offset! @injection.content 0 1 0 -1)
>   (#set! injection.include-children)
> )

You don't need `#offset!` if you inject into `string_content`.

lewis6991 commented:
Sure, just mentioning, so there's a trace this exists on the internet. Also in case someone raises a shebang PR.

lewis6991 commented:
> You don't need #offset! if you inject into string_content.

Can't use `string_content` when you have this:

```
      argument: (string ; [6, 5] - [16, 1]
        (string_content) ; [7, 0] - [7, 4]
        (string_content) ; [8, 0] - [8, 23]
        (string_content) ; [9, 0] - [9, 11]
        (string_content) ; [10, 0] - [10, 19]
        (string_content) ; [11, 0] - [11, 19]
        (string_content) ; [12, 0] - [12, 14]
        (command_substitution ; [12, 14] - [12, 20]
          (command ; [12, 16] - [12, 19]
            name: (command_name ; [12, 16] - [12, 19]
              (word)))) ; [12, 16] - [12, 19]
        (string_content) ; [12, 20] - [12, 27]
        (string_content) ; [13, 0] - [13, 16]
        (string_content) ; [14, 0] - [14, 18]
        (string_content))) ; [15, 0] - [15, 42]
```

Unless we use `injection.combined`, but I don't think we want that.

clason committed:
feat(treesitter): add `@injection.filename`

Problem: Injecting languages for file redirects (e.g., in bash) is not
possible.

Solution: Add `@injection.filename` capture that is piped through
`vim.filetype.match({ filename = node_text })`; the resulting filetype
(if not `nil`) is then resolved as a language (either directly or
through the list maintained via `vim.treesitter.language.register()`).

Note: `@injection.filename` is a non-standard capture introduced by
Helix; having two editors implement it makes it likely to be upstreamed.

clason commented:
I don't feel comfortable adding it to queries in this case; I also don't like deviating from upstream (nvim-treesitter). Let's let those cook out of core for a while.

lewis6991 commented:
Huh? What's so uncomfortable about `offset` and what's "deviating"?
